src/Stratis.Features.SQLiteWalletRepository/External/ScriptDestinationReader.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using NBitcoin;
+using NBitcoin.DataEncoders;
+using Stratis.Bitcoin.Interfaces;
+
+namespace Stratis.Features.SQLiteWalletRepository.External
+{
+    public interface IScriptDestinationReader : IScriptAddressReader
+    {
+        IEnumerable<TxDestination> GetDestinationFromScriptPubKey(Network network, Script script);
+    }
+
+    public class ScriptDestinationReader : IScriptAddressReader
+    {
+        private readonly IScriptAddressReader scriptAddressReader;
+
+        public ScriptDestinationReader(IScriptAddressReader scriptAddressReader)
+        {
+            this.scriptAddressReader = scriptAddressReader;
+        }
+
+        public string GetAddressFromScriptPubKey(Network network, Script script)
+        {
+            return this.scriptAddressReader.GetAddressFromScriptPubKey(network, script);
+        }
+
+        public virtual IEnumerable<TxDestination> GetDestinationFromScriptPubKey(Network network, Script redeemScript)
+        {
+            string base58 = this.scriptAddressReader.GetAddressFromScriptPubKey(network, redeemScript);
+
+            if (base58 != null)
+            {
+                TxDestination destination = ScriptDestinationReader.GetDestinationForAddress(base58, network);
+
+                if (destination != null)
+                    yield return destination;
+            }
+        }
+
+        public static TxDestination GetDestinationForAddress(string address, Network network)
+        {
+            if (address == null)
+                return null;
+
+            byte[] decoded = Encoders.Base58Check.DecodeData(address);
+            return new KeyId(new uint160(decoded.Skip(network.GetVersionBytes(Base58Type.PUBKEY_ADDRESS, true).Length).ToArray()));
```

Comment:
I'm not entirely sure I understand the purpose of this. You appear to be using it in one other place (`GetDestinations`), but in that place you're converting straight back to a scriptPubKey?

Reply:
We have the situation where the existing `ScriptAddressReader` has some limitations: (1) It only returns one address, and (2) it returns an address where a ScriptPubKey, for matching directly with the `HDAddress`'s `ScriptPubKey` may be preferable. To get around this I implemented `ScriptDestinationReader` that for the time being leverages any logic implemented under the old (current) `ScriptAddressReader`.